spec/users_controller_spec.rb

```diff
@@ -0,0 +1,96 @@
+require_relative 'spec_helper'
+
+def app
+  UsersController.new
+end
+
+describe UsersController do
+  before do
+    @users = DB[:users]
+  end
+ 
+  #-------------------------------------------------------------------------------------
+  describe "when logged in as an admin" do
+    before do
+      @admin = @users.where(admin: true).first
+      env "rack.session", {:current_user_id => @admin[:id]}
+    end
+
+    it "able to access '/users' page" do
+      get '/users'
+      last_response.body.must_include "Complete CRUD Users"
+    end
+
+    it "able to access other user's page" do
+      different_user_id = @admin[:id] + 1
+      get "/users/#{different_user_id}"
+      last_response.body.must_include "User Info"
+    end
+
+    it "displays all users '/users' page" do
+      number_of_users = @users.count
+      get '/users'
+      users_page = Nokogiri::HTML(last_response.body)
+      li_size = users_page.search("li").size
+      li_size.must_equal number_of_users
+    end
+  end
+
+  #-------------------------------------------------------------------------------------
+  describe "when logged in as a regular user" do
+    before do
+      @regular_user = @users.where(admin: false).first
+      env "rack.session", {:current_user_id => @regular_user[:id]}
+    end
+
+    it "can view own page" do
+      get "/users/#{@regular_user[:id]}"
+      last_response.body.must_include "User Info"
+    end
+
+    it "redirects regular_user when accessing '/users' page" do
+      get '/users'
+      follow_redirect!
+      last_response.body.must_include "Unauthorized access"
+    end
+
+    it "redirects regular_user when accessing other user's page" do
+      different_user_id = @regular_user[:id] + 1    
+      get "/users/#{different_user_id}"
+      follow_redirect!
+      last_response.body.must_include "Unauthorized access"
+    end
+  end
+
+  #-------------------------------------------------------------------------------------
+  describe "testing login, create, update and delete" do
+    before do
+      post '/users', params = {fname: "f3", lname: "l3", email: "e3", username: "u3", password: "p3"}
+      @current_user = @users.where(fname: "f3").first
```

Comment:
Let's be consistent with the use of the term `current_user`. Best to use it only when referring to a signed-in user. In this case, better to stick with just `user`.